amazon-kinesis-client/src/main/java/software/amazon/kinesis/retrieval/fanout/FanOutRecordsPublisher.java

```diff
@@ -313,8 +319,24 @@ private void errorOccurred(RecordFlow triggeringFlow, Throwable t) {
         }
     }
 
-    private void clearRecordsDeliveryQueue() {
+    // This method is not thread safe. This needs to be executed after acquiring lock on this.lockObject
+    private void clearRecordsDeliveryQueue(boolean isSubscriptionCompleting) {
+        if(isSubscriptionCompleting) {
+            // This will prevent further events from getting scheduled
+            isAwaitingFinalAckForCurrentSubscription = true;
+            try {
+                if (!recordsDeliveryQueue.isEmpty()) {
+                    // Wait for the configured time to get a notification for already delivered event, if any.
+                    lockObject.wait(TIME_TO_WAIT_FOR_FINAL_ACK_MILLIS);
```

Comment:
Why are we not waiting if the delivery queue is empty. There might still be a event in which was already delivered?

Reply:
when the delivery queue is empty then it is guaranteed that there will be no valid in-flight event.